core/src/saros/concurrent/jupiter/internal/Jupiter.java

```diff
@@ -142,22 +144,17 @@ public boolean isCurrent(Timestamp timestamp) throws TransformationException {
   public Operation receiveJupiterActivity(JupiterActivity jupiterActivity)
       throws TransformationException {
 
-    if (jupiterActivity.getOperation() instanceof TimestampOperation) {
-
-      // TODO Use timestamps correctly!
-      log.warn("Timestamp operations are not tested at the moment");
-
-      updateVectorTime(jupiterActivity.getTimestamp());
```

Comment:
This was the last usage of the method `updateVectorTime`. Sadly, as the mehtod does not have any meaningful javadoc (and the commit message when it was added is not helpful either), I am not sure what exactly it was meant for (why it was added). It seems like it was meant to be used together with the TimestampOperation to remotely update the vector-time of other participants. But why this was added is not clear to me.

I am somewhat torn on this, but I would prefer not to do half measures here. In my opinion, we should either
- completely remove the update stuff (as it is not used anywhere) if you want to reuse the TimestampOperation for heartbeats (I still think the name could be misleading) or
- create a new activity/operation for the heartbeat and leave the TimestampOperation stuff untouched (for now).

I am currently leaning towards the second option (implementing a new operation with a more fitting name) as this gives us the opportunity to evaluate the update logic more thoroughly without blocking this PR. Furthermore, I am not sure whether the functionality to update remote timestamps could still be useful for solving #223.

Reply:
Things I would like to see:

A dedicated class / component that can be turned on/off and not some thread hidden inside the client class doing "magic heartbeats".

The heartbeat firing should be done in the EDT. It is possible in this setup that you will fire a heartbeat for a non existing file (Eclipse filesystem is concurrent).

Last but not least what is wrong with generating NOPs ? They should also do the trick. Of course this ain't not so easy because you have to fire special TextActivities that will be translated to NOPs.

Reply:
@tobous Because `TimestampOperation` was not used anywhere meaningful, I just went with it. But sure, I can add another operation, if you want to keep the code for #223. I will add a note on its current state.

@srossbach 

> Last but not least what is wrong with generating NOPs ?

The reason I need a new operation type and not a NOP is [this line](https://github.com/saros-project/saros/pull/397/files#diff-8d4e1fd03035f8adc5945d18e4a6f68aR85). NOPs would get added to the `ackJupiterActivityList`, which is absolutely unnecessary for the heartbeat and might cause more harm then this small special case does.

> A dedicated class / component that can be turned on/off

Sure I can move this into a separate class (and make it togglable). Are you fine with spawning the instance in the `JupiterClient`, because it needs access to the `clientDocs`?

> .. EDT ...
What is that?

> (Eclipse filesystem is concurrent)
In that case the `JupiterClient` will be out-of-sync anyway, which does not seem to be a problem currently, so is there something I need to do about this? This might actually be of use because I am not sure if the stored operations would be cleared if a file is closed. So a heartbeat for a missing file might actually be desirable.

Reply:
EDT = Event Dispatch Thread aka the Main / GUI thread of an application.

Reply:
To execute code on the IDE specific EDT, we have the core interface [UISynchronizer](https://github.com/saros-project/saros/blob/master/core/src/saros/synchronize/UISynchronizer.java).

(Somewhat off-topic: @srossbach I just noticed that we still have an old implementation for an AWT synchronizer located in the eclipse code that is never used: [AWTSynchronizer](https://github.com/saros-project/saros/blob/master/eclipse/src/saros/synchronize/internal/AWTSynchronizer.java). The commit message suggests that it was added for compatibility's sake. I don't think it is needed anymore as this is now handled separately by the IDE implementations. Am I correct in this assumption?)